### PR TITLE
Refactoring `git` domain to use async generators, remove readable-stream, simplify utilities

### LINKED ISSUE #298: Review and Refactor of `git/utils.ts` for Serving Vault Git Repositories
### Specification

The underlying workings of `git/utils.ts` is largely a mystery to our team. We want to perform a deep dive into the workings of this code, and remove/refactor any of our own implementation, potentially also replacing their usage in favour of the `isomorphic-git/internal-apis`. There seems to be some overlap with what our git utils perform, and what `iso-git` provides.

To clarify... we use git in 2 ways: as a client and as a server.

When using isogit, it only provides the primitives to use git as a client. It does not provide the primitives to use git as a server.

Why do we use git as a client and server? Well our entire secrets management system is built as a stack, from bottom to top:

1. Encrypted Key Value Database (soon to have transparent block based encryption) - js-db
2. Virtual Filesystem - js-encryptedfs
3. Virtual Git plus Filesystem - isogit (this is what constitutes a "vault")
4. Virtual Unix Operations - MatrixAI/Polykey-CLI#32 (this is what creates "vault operations")
5. Frontends - Polykey-CLI, Polykey-Desktop, Polykey-Mobile (and then presented over an RPC interface to different front-ends)

As you can see, the "Virtual Git plus Filesystem" layer makes use of isogit. A virtual git repository is created on the virtual filesystem. This is how we are able to then stack unix operations with roughly similar semantics later, in order to maintain API consistency with what developers are used to.

Now the issue is that isogit by itself can act like a git client. That means it can clone/pull repositories, it can create repositories, it can create commits, it can manage the git objects... etc.

However isogit cannot "serve" the repositories. And this is essential to be able to share our vaults.

To serve our repositories we need to know how to act like a "git server" when we use isogit or something were to use git to try and clone/pull our vaults.

So a long time ago, we developed the git server primitives and put them all into `git/utils.ts`. However this code has not been heavily tested, nor has it gone through any refactoring since then. And since then we have forgotten how it all works, and it's not in the programming style that we have developed over the last few years working on Polykey.

So the `git/utils.ts` needs to go under review, and most likely some heavy refactoring with the addition of tests, to ensure that acting like a git server makes sense here.

Now there is definitely cross over between isogit and the `git/utils.ts`. However we aren't aware of what these are. And there probably should be a bunch of "shared types" and similar constructs. We just need to get into it, and get it done.

### How this was addressed

After reviewing the code as it was, we ended up just cleaning it up by refactoring the `http` response to be streamed using an async generator. Most of the utility functions were removed in favour of using the `isomorphic-git` plumbing functions for interacting with the git data structure.

Ultimately this didn't speed things up very much, however, it did significantly reduce the complexity of the git domain. So it is much easier to follow along with what is happening. There is also a slight improvement with pulling a repo. We use the `haves` provided by the client to only grab the minimum objects required for the pull.

### Additional context

* initial discussion of git utils in sprint 11 meeting (roughly around the 1:02:00 mark): https://vimeo.com/manage/videos/650965464

### Tasks

1. Review the whole git domain
2. Refactor the git domain with the intention of simplifying the code and improving performance when cloning/pulling a vault.

### LINKED ISSUE #523: Replace `readable-stream` usage with Web Streams in the Vaults and Git Domains
### Specification

Some parts of the code are still using `readable-stream`. these need to be replaced with webstreams and the `readable-stream` dependency removed from the `package.json`.

These usages can be found at 
```

[nix-shell:~/Projects/Polykey]$ ag "from 'readable"
src/vaults/VaultManager.ts
24:import { PassThrough } from 'readable-stream';

src/git/types.ts
1:import type { PassThrough } from 'readable-stream';

src/git/utils.ts
22:import { PassThrough } from 'readable-stream';
```

### Additional context

* https://github.com/MatrixAI/js-encryptedfs/issues/75 - Git related streams may have some relationship with EFS readable stream changes, can also fix that up too (however EFS will cannot use web streams, since it has to simulate node FS streams)

### Tasks

1. ~[ ] 1. Replace any `readable-stream` usage with `stream/web`.~ - replacing with async generator
2. [x] 2. Remove `readable-stream` as a dependency from the `package.json`.

### PR DESCRIPTION
**Pull request by tegefaulkes**
### Description
This PR addresses refactoring the `git` domain to be more clear and more in line with our latest practices.

### Issues Fixed

* Fixes #298 
* Ref ENG-68
* Fixes #523 

### Tasks

- [x] 1. Factor out http stream construction to a new `http.ts` file.
- [x] 2. Keep git and FS interaction utility in `utils.ts`.
- [x] 3. Refactor existing code that we are keeping to be more in line with current practices.
- [x] 4. Refactor the HTTP response to be streaming focused. Using generators for most of the back end logic and convert to a webstream for the rpc stage.
- [x] 5. All types should be formalised where appropriate.
- [x] 6. Inline documentation should be updated and all relevant spec details should be included along side the code.
- [x] 7. Tests need to updated and refactored to reflect the new refactoring changes.
- ~8. Performance is an issue here, so benchmarks should be added to test for future regressions.~ - Differed to a new PR. It will test the current changes but nothing here will depend on benchmarks, so I'm just going to make it separate.
- [x] Check if #523 has been addressed fully

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

### CODE REVIEW

**Comment by tegefaulkes** [comment]
The reference discovery stage has been completed. I implemented it as `AsyncGenerator`s which made it stream-able but also very clean logic wise. I opted for generators over `webstreams` for this stage since composing streams together is a lot cleaner with generators than streams.

**Comment by CMCDragonkai** [comment]
Generators are a higher level construct indeed. I think streams represent a lower level construct, and also streams are actually synchronous. Whereas generators represent coroutines in the javascript runtime. So if you're dealing at the object-level then async generator abstraction is better abstraction.

**Comment by CMCDragonkai** [comment]
In the future, you should write up your own branch name.

**Comment by CMCDragonkai** [comment]
New types and new classes need to be documented into the spec. Like this:

#### Module

```
type X = ...
```

```
type Y = ...
```

Give a reason as why they exist now.

File: src/git/http.ts [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+ * S: Content-Type: application/x-git-upload-pack-advertisement
+ * S: Cache-Control: no-cache
+ * S:
+ * S: 001e# service=git-upload-pack\n
+ * S: 0000
+ * S: 004895dcfa3633004da0049d3d0fa03f80589cbcaf31 refs/heads/maint\0multi_ack\n
+ * S: 003fd049f6c27a2244e12041955e262a404c7faba355 refs/heads/master\n
+ * S: 003c2cb58b79488a98d2721cea644875a8dd0026b115 refs/tags/v1.0\n
+ * S: 003fa3c2e2402b99163d1d59756e5f207ae21cccba4c refs/tags/v1.0^{}\n
+ * S: 0000
```
Comment (CMCDragonkai) [inline src/git/http.ts:66]: Beware of the parser generator standard.

Comment (CMCDragonkai) [reply src/git/http.ts:66]: There's examples of that in the `ids`. And there's an old issue describing the whole concept.

Comment (CMCDragonkai) [reply src/git/http.ts:66]: Still to align with that standard.

That information needs to be put into the development docs.

Polykey-Docs - MatrixAI-Graph.

Comment (CMCDragonkai) [reply src/git/http.ts:66]: The `claims` domain also uses it.

**Comment by CMCDragonkai** [comment]
Expected to take the whole cycle.

**Comment by CMCDragonkai** [comment]
image

This is a bug with the vault log - pull system. Needs to be fast forward merge.

**Comment by tegefaulkes** [comment]
Progress update: I've converted to using generators and swapped over to using them in the code. The test I've been using as a sanity check is passing as well. I'm going to go back through the code and clean it up a little now.

**Comment by tegefaulkes** [comment]
`isometric-git` actually provides a lot of the functionality we needed in the git domain. I was actually able to replace most of the `utils.ts` functions with it. I did some cursory checks and the things run just as fast, the whole git domain has been simplified immensely.

It seems odd that `isometic-git` wasn't used to handle interacting with the git structure in the first place. Was the git domain created before we started using `isometric-git`?

**Comment by CMCDragonkai** [comment]
Originally isogit did not have server-side functionality. Only client side. You were able to replace it all? I thought there was git server functionality that was missing. Remember in the future we may want to expose an http API.

**Comment by tegefaulkes** [comment]
It is missing the server side functionality, but it does provide a lot of plumbing functionality for interacting with the git structure that the `utils.ts` was replicating.

For example, Listing all the existing refs, resolving them to objectIds, then finding all of the dependent objectIds, and finally creating the packfile format that gets sent over can all be done with isogit. The only thing I had to do was generate the HTTP responses in the proper format.

**Comment by CMCDragonkai** [comment]
Then yes it was a quick and dirty job.

**Comment by CMCDragonkai** [comment]
Originally we even wanted to use libgit2 but it was easily adapted to work on a in-memory virtual filesystem. Too tied into posix fs semantics.

**Comment by tegefaulkes** [comment]
Mostly done now. Just need to fix some tests in the vaults domain now. Along with doing the benchmarks.

I think the benchmarks can be moved to a 2nd stage PR. I also think the whole vaults domain needs a general review and touch up, also in a separate PR.

**Comment by tegefaulkes** [comment]
Tests are fixed, I added a end-to-end integration tests of of cloning and pulling in the `http.test.ts` tests. This does a git clone/pull using just the new `http` code directly and the normal file system. So no `efs` or networking. It should be a good basis of comparison between adding `efs` and networking separately in the benchmarks. 

I'm going to look into the clone/pull bug mentioned from before and then move on the merge prep. I'll start a new PR for adding the benchmarks.

**Comment by tegefaulkes** [comment]
1 test relating to discovery and 2 tests relating to MDNS are consistently failing. I'm not sure why and they're wait out of scope of this work so it shouldn't be affected.

**Comment by tegefaulkes** [comment]
This is ready for review. I'm going to start the benchmarks in a new PR for now.

File: src/git/http.ts (hunk 1 above)
Comment (CMCDragonkai) [reply src/git/http.ts:66]: What's the status with this?

File: tests/git/utils.ts [hunk 2]
```diff
@@ -1,72 +1,270 @@
...
+import type fs from 'fs';
 import path from 'path';
 import git from 'isomorphic-git';
+import fc from 'fast-check';
+import * as gitUtils from '@/git/utils';
+import * as gitHttp from '@/git/http';
+import * as utils from '@/utils';
 
+// Just to avoid confusing the type with the name
+type FsType = typeof fs;
```
Comment (CMCDragonkai) [inline tests/git/utils.ts]: Can we use the filesystem type?

There's already a `import type fs from 'fs'` in `src/types.ts` in Polykey. Probably one in PK CLI too?

File: tests/git/utils.ts [hunk 3]
```diff
@@ -1,72 +1,270 @@
...
+    }
+  | {
+      type: 'have';
+      objectId: string;
+    }
+  | {
+      type: 'SEPARATOR' | 'done' | 'none';
+    };
+
+function generateTestNegotiationLine(data: NegotiationTestData, rest: Buffer) {
```
Comment (CMCDragonkai) [inline tests/git/utils.ts]: Can we get some docblocks for this. Also `generateTest...` sounds a bit weird?

File: tests/git/utils.ts [hunk 4]
```diff
@@ -1,72 +1,270 @@
...
+      return Buffer.concat([Buffer.from('0009done\n'), rest]);
+    case 'none':
+      return Buffer.alloc(0);
+    default:
+      utils.never();
+  }
+}
+
+// Used to print out the contents of an `Buffer` iterable for testing
+async function* tapGen(
```
Comment (CMCDragonkai) [inline tests/git/utils.ts]: Docblock format please.

Comment (CMCDragonkai) [reply tests/git/utils.ts]: This function's name is a bit weird? Should it be expanded.

File: tests/git/utils.ts [hunk 5]
```diff
@@ -1,72 +1,270 @@
...
+  gen: AsyncIterable<Buffer>,
+): AsyncGenerator<Buffer, void, void> {
+  let acc = '';
+  for await (const line of gen) {
+    acc += line.toString();
+    yield line;
+  }
+  // eslint-disable-next-line no-console
+  console.log(acc);
```
Comment (CMCDragonkai) [inline tests/git/utils.ts]: No `console.log` allowed.

File: tests/git/utils.ts [hunk 6]
```diff
@@ -1,72 +1,270 @@
...
+  listGitObjects,
+  generateTestNegotiationLine,
+  tapGen,
+  request,
+  objectIdArb,
+  capabilityArb,
+  capabilityListArb,
+  restArb,
+  wantArb,
+  haveArb,
```
Comment (CMCDragonkai) [inline tests/git/utils.ts]: I think all these names should be expanded a bit. Like `wantArb`? That's way too simple. It should be specific to particular data structures within git.

File: tests/git/utils.ts [hunk 7]
```diff
@@ -1,72 +1,270 @@
...
+  lineDataArb,
```
Comment (CMCDragonkai) [inline tests/git/utils.ts]: This too!

Comment (CMCDragonkai) [reply tests/git/utils.ts]: I would suggest prefixing it like `gitXYZArb`.

File: src/git/http.ts [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+    yield packetLineBuffer(subBuffer, gitUtils.CHANNEL_DATA);
+  } while (packFileBuffer.byteLength > chunkSize);
+}
+
+export {
+  advertiseRefGenerator,
+  packetLineBuffer,
+  parsePackRequest,
+  generatePackRequest,
+  generatePackData,
```
Comment (CMCDragonkai) [inline src/git/http.ts:424]: I would like docblocks here to make it clear exactly the dependency tree, or call graph of the http protocol itself.

As in what is considered the main function in relation to the cloning and pulling operations? And how does it decompose subsequently. Docblocks should use ASCII diagrams to make this clear.

**Comment by CMCDragonkai** [comment]
I added some additional review comments with a first glance. Also this review comment previously should be addressed: https://github.com/MatrixAI/Polykey/pull/709#pullrequestreview-2039845891.

Otherwise is there specific things you want reviewed?

**Comment by CMCDragonkai** [comment]
The final checklist should be ticked off.

**Comment by tegefaulkes** [comment]
Nothing specific needs to be reviewed, just need the extra pair of eyes for the due diligence.

I'm leaving the checklist along for when the review changes have been addressed.

File: src/git/http.ts (hunk 1 above)
Comment (tegefaulkes) [reply src/git/http.ts:66]: I modified the parsers to be consistent with how the claims domain does things. I ended up creating an `isX()` and `assertX()` to use.

File: tests/git/utils.ts (hunk 2 above)
Comment (tegefaulkes) [reply tests/git/utils.ts]: It makes more sense to use `FileSystemWritable` or `EncryptedFs` as the type here. `isomorphic-git` doesn't like the `FileSystem` type for reasons I haven't dug into.

I think I'll just make sure everything that uses the `fs` works with the `EFS` and for testing we use `fs` but override the type where needed.

Comment (tegefaulkes) [reply tests/git/utils.ts]: Switched everything relevant to `efs: EncryptedFs`. This should make the domain more consistent.

File: tests/git/utils.ts (hunk 3 above)
Comment (tegefaulkes) [reply tests/git/utils.ts]: Added some clarification for it.

File: tests/git/utils.ts (hunk 5 above)
Comment (tegefaulkes) [reply tests/git/utils.ts]: If we're that strict about the `console.log` then I'll just remove the `tapGen` function. It's only there for testing and not even used.

File: tests/git/utils.ts (hunk 4 above)
Comment (tegefaulkes) [reply tests/git/utils.ts]: `tapGen` has been removed.

File: tests/git/utils.ts (hunk 5 above)
Comment (tegefaulkes) [reply tests/git/utils.ts]: `tapGen` has been removed.

File: tests/git/utils.ts (hunk 6 above)
Comment (tegefaulkes) [reply tests/git/utils.ts]: I have cleaned these up. Added some comments and made the names more specific.

File: tests/git/utils.ts (hunk 7 above)
Comment (tegefaulkes) [reply tests/git/utils.ts]: Fixed along with above.

File: src/git/http.ts (hunk 8 above)
Comment (tegefaulkes) [reply src/git/http.ts:424]: Updated the relevant docblocks.

**Comment by tegefaulkes** [comment]
Done and ready to merge. I'm going to rebase the related PRs before merging this.

**Pull request merged**